### ISSUE TITLE
Enhancement: make lint() accept string input

### ISSUE THREAD

**Issue by fangly** [jimhester/lintr#234]
Jim suggested improving the `lint()` function by making it accept a character **string** through the 'filename' option: https://github.com/jimhester/lintr/pull/233#issuecomment-297761852 . `lint()` would work in string mode if 'filename' contains a trailing newline character.

```R
lintr::lint("a = b\n")
```

It seems to be partially implemented in `lint()`; the code seems to be there, but as far as I can tell is not tested nor documented.

The 'filename' argument does not seem ideal for this purpose, but if we found a better alternative, we would need to deprecate 'filename', which is also not ideal. Besides, 'filename' is the first argument and can therefore be omitted in the majority of cases.

In my eyes, the main issue with this enhancement proposal is that it makes it impossible to pass strings without newlines, and therefore to test the trailing newline linter. Perhaps the requirements should be relaxed and a newline character should be detected anywhere in the string?
```R
lintr::lint("NULL\na = b")
```

Alternatively, string input may be passed in a list object containing lines of text.

```R
lintr::lint(list("a = b"))
```

After implementing this functionality, the `expect_lint()` function could be simplified and pass its arguments directly to `lint()`.

**Comment by jimhester** [comment]
The code is fine how it is, the last newline is just removed, so to test the trailing newline linter you just use two newlines at the end.

```r
lintr::lint("trailing_newline\n\n")
#> <text>:2:1: style: Trailing blank lines are superfluous.
#> 
#> ^
```

**Comment by russHyde** [comment]
Is this feature implemented? I can't get @jimhester 's `lintr::lint("trailing_newline\n\n")` to parse using the current master of lintr. 

Traceback:
```
library(lintr)
> lintr::lint("trailing_newline\n\n")
Error in file(con, "r") : cannot open the connection
In addition: Warning message:
In file(con, "r") : cannot open file 'trailing_newline

': No such file or directory
> traceback()
4: file(con, "r")
3: readLines(filename)
2: get_source_expressions(filename)
1: lintr::lint("trailing_newline\n\n")
```

**Comment by kenahoo** [comment]
This feature does not seem to work anymore:

```
> packageVersion('lintr')
[1] ‘1.0.3’

> lintr::lint("5+5\n")
Error in file(con, "r") : cannot open the connection
In addition: Warning message:
In file(con, "r") : cannot open file '5+5
': No such file or directory
```

Is there another way to test a linter directly against a string?

**Comment by jimhester** [comment]
Works fine, it is just not in the CRAN release.
